### PR TITLE
Mutex protection for db writing and stl collections in writer & storage

### PR DESCRIPTION
**Pull request by adamdbrw**
Adresses the https://github.com/ros2/rosbag2/issues/602 issue.

The issue is with that the current locking mechanism on top level (mutexes in writer.cpp in rosbag2_cpp, added with https://github.com/ros2/rosbag2/pull/491) is valid only with assumption that all writes of data and all creation/removal of topics are synchronous. This is not the case since double buffering has been introduced in the writer. This also would not be the case if another, non-sequential writer implementation was added, and would be unnecessary for another, parallel-enabled storage implementation.

This PR however does not remove mutexing on the rosbag2_cpp level, only adds storage implementation level locks - lets discuss/test whether remaining (intermittent) layers are multithreading-ready.

### CODE REVIEW

**Comment by adamdbrw** [comment]
The failing test (due to a timeout) for reading looks unrelated and the same test seems to be failing on master as well

**Comment by mjeronimo** [comment]
My understanding so far:
* The mutex usage at the rosbag2 writer level are only valid in the case that the back-end implementation is synchronous (it basically assumes a synchronous implementation). 
* Currently, the sqlite3 back-end opens sqlite3 databases using SQLITE_OPEN_NOMUTEX, which means that separate threads are allowed to use SQLite at the same time, as long as each thread is using a different database connection. However, since we are seeing this crash, I'd assume that it is not true that each thread is using a different database connection.
* The proposed solution keeps the above while introducing another mutex at the Writer level of the sqlite3 back-end, used for any database write operations, effectively compensating for multiple threads using the same database connection. 

IMO,
* The mutex at the rosbag2 writer level should be removed and the burden of thread safety should be on the back-end implementation. 
* Regarding the sqlite3 back-end, if performance does not seem to be an issue, why not push the responsibility further by opening the databases with SQLITE_OPEN_FULLMUTEX and not having a mutex at the Writer level? Then, the responsibility is only in sqlite3 (versus what would be in three places). Vectors of messages seemed to be a potential concern, but it isn't clear if there is a performance issue in practice (because of the I/O dominating).

**Comment by adamdbrw** [comment]
Ok this is ready to be merged. I tested that no locking occurs as well with version (in a case that was apparently deterministically causing it before this PR). @mjeronimo.

**Comment by adamdbrw** [comment]
@csli8192 this fix does not include discovery mechanism changes which you earlier suggested. If they are still necessary for your use case, I believe is best to merge them in separately.

**Comment by mjeronimo** [comment]
@adamdbrw Thanks for the detailed explanation. I now understand that the intermediate data structures also have to be protected and that it's not enough to just use FULLMUTEX.

**Review by mjeronimo — APPROVED** [review]

File: rosbag2_storage_default_plugins/include/rosbag2_storage_default_plugins/sqlite/sqlite_storage.hpp [hunk 1]
```diff
@@ -92,6 +92,7 @@ class ROSBAG2_STORAGE_DEFAULT_PLUGINS_PUBLIC SqliteStorage
   void fill_topics_and_types();
   void activate_transaction();
   void commit_transaction();
+  void write_locked(std::shared_ptr<const rosbag2_storage::SerializedBagMessage> message);
```
Comment (emersonknapp) [inline rosbag2_storage_default_plugins/include/rosbag2_storage_default_plugins/sqlite/sqlite_storage.hpp:95]: Where is this function used? Since it's private I don't see how it can connect to any existing code, so I'm confused

File: rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_wrapper.cpp [hunk 2]
```diff
@@ -42,7 +42,7 @@ SqliteWrapper::SqliteWrapper(
   if (io_flag == rosbag2_storage::storage_interfaces::IOFlag::READ_ONLY) {
     int rc = sqlite3_open_v2(
       uri.c_str(), &db_ptr,
-      SQLITE_OPEN_READONLY | SQLITE_OPEN_NOMUTEX, nullptr);
+      SQLITE_OPEN_READONLY | SQLITE_OPEN_FULLMUTEX, nullptr);
```
Comment (emersonknapp) [inline rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_wrapper.cpp:45]: Is there a regression test to verify that this changes any measurable behavior?

**Comment by adamdbrw** [comment]
We have given it a deeper thought and ran extensive tests. The race condition was still apparently there (in reality no, a separate concurrency problem), and the version 1 of this changeset was the correct one. Why?

SqliteStorage has unordered_map structure holding the topics. It is not a thread safe structure. As it is now, create_topic can be run from one thread and initiate `.emplace`. Meanwhile, the buffer writer thread will attempt to find a topic for the message it is trying to insert, running `.find` on the same structure. As you likely know, this threading scenario is not supported in stl containers.

In particular, for unordered_map, a likely case is rehashing. "If rehashing occurs due to the insertion, all iterators are invalidated. Otherwise iterators are not affected". 

Thus a rare case which we observed in extensive testing with over a thousand of topics (thanks @pijaro). 

Anyway, bottom line is that any stl collections that can be accessed outside of supported model (only 1 thread writes while 0 read, or any number of threads read) should be locked. We don't need to dig deep into details such as rehashing during find - we just need to lock them. And with locking on a member function level, we don't need FULLMUTEX anymore - it is completely redundant.

File: rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_wrapper.cpp (hunk 2 above)
Comment (adamdbrw) [reply rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_wrapper.cpp]: This is not there anymore - but yes, a good catch since it wasn't needed in READ_ONLY

File: rosbag2_storage_default_plugins/include/rosbag2_storage_default_plugins/sqlite/sqlite_storage.hpp [hunk 3]
```diff
@@ -92,6 +93,7 @@ class ROSBAG2_STORAGE_DEFAULT_PLUGINS_PUBLIC SqliteStorage
...
```
Comment (emersonknapp) [inline rosbag2_storage_default_plugins/include/rosbag2_storage_default_plugins/sqlite/sqlite_storage.hpp:96]:
```suggestion
  void write_locked(std::shared_ptr<const rosbag2_storage::SerializedBagMessage> message) REQUIRES(write_mutex_);
```

**Review by emersonknapp — APPROVED** [review]
This looks good - I'm going to (optionally) ask for Clang thread safety annotations, as they will help validate and clarify the logic a little bit.

e.g. https://github.com/ros2/rcpputils/blob/master/include/rcpputils/thread_safety_annotations.hpp#L133

I'm not sure if you need this across the class - but if so, you could guard the two class members that you seem to be synchronizing access to

```
std::unordered_map<std::string, int> topics_ GUARDED_BY(database_write_mutex_);
std::shared_ptr<SqliteWrapper> database_ GUARDED_BY(database_write_mutex_);
```

File: rosbag2_storage_default_plugins/include/rosbag2_storage_default_plugins/sqlite/sqlite_storage.hpp (hunk 3 above)
Comment (adamdbrw) [reply rosbag2_storage_default_plugins/include/rosbag2_storage_default_plugins/sqlite/sqlite_storage.hpp]: Done - thank you for suggestions! I didn't know we had such annotations. 
I also rebased this PR on the new master.

**Review by emersonknapp — APPROVED** [review]

**Comment by emersonknapp** [comment]
Gist: https://gist.githubusercontent.com/emersonknapp/ca5fc9b7640834b2ec160af39e38a1c2/raw/c00c5de4bbf391319d623fde04f97a2a0200aa1b/ros2.repos
BUILD args: --packages-up-to rosbag2_storage_default_plugins rosbag2_tests
TEST args: --packages-select rosbag2_storage_default_plugins rosbag2_tests
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/7502
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13481/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8389/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11202/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13546/)

**Comment by adamdbrw** [comment]
All looks green, shall we merge this? :]
